tool/tsh/db.go

```diff
@@ -727,15 +720,21 @@ func onDatabaseConnect(cf *CLIConf) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	routeToDatabase, database, err := getDatabaseInfo(cf, tc, cf.DatabaseService)
+	route, database, err := getDatabaseInfo(cf, tc, cf.DatabaseService)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	if routeToDatabase.Protocol == defaults.ProtocolDynamoDB {
-		return trace.BadParameter(formatDbCmdUnsupportedDBProtocol(cf, routeToDatabase))
+	if route.Protocol == defaults.ProtocolDynamoDB {
+		return trace.BadParameter(formatDbCmdUnsupportedDBProtocol(cf, route))
 	}
-	if err := maybeDatabaseLogin(cf, tc, profile, routeToDatabase); err != nil {
-		return trace.Wrap(err)
+
+	pref := getDBLocalProxyPreference(cf.Context, tc, route)
+	if !(pref.localProxy && pref.tunnel) {
+		// We only need to (possibly) login if not using a local proxy tunnel,
+		// because a local proxy tunnel will handle db login itself.
+		if err := maybeDatabaseLogin(cf, tc, profile, route); err != nil {
```

Comment:
`maybeDatabaseLogin` is shared by `tsh db connect` and `tsh proxy db`, so a weakness of the way I've written this is that `tsh proxy db` does not respect the "preferred" tunnel setup - it will only tunnel if `--tunnel` flag is given. Therefore if I move `pref.localProxy && pref.tunnel` check into `maybeDatabaseLogin` it will need more info to know whether the preference will actually be used... `tsh db connect` uses the preference so it can skip login, `tsh proxy db` only looks at the local proxy requirements, so it still needs to login when tunnel is preferred but not required.

It's complicated 😅 I can refactor `maybeDatabaseLogin` to add a preference arg and "uses preference" bool arg (or something)

Reply:
then don't worry about it =D

Reply:
refactored to pass requirements to `maybeDatabaseLogin` and eliminated the "preferences" struct altogether. I think this simplified things a lot